### ISSUE TITLE
[Custom Order Tables] Failed payment when processing a subscription renewal

### ISSUE THREAD

**Issue by mattallan** [Automattic/woocommerce-payments#4548]
⚠️ This issue will be fixed in `subscriptions-core` library. Track progress on the issue here: https://github.com/Automattic/woocommerce-subscriptions-core/issues/146

---

### Describe the bug

Custom Order Tables are coming in WC 6.9 and while testing Subscriptions with WCPay I noticed that processing a subscription renewal results in the subscription going on-hold and the renewal order has order note "Failed Payment".

In the debug.log there's two additions:

```
[09-Aug-2022 06:32:51 UTC] PHP Notice:  Function is_internal_meta_key was called <strong>incorrectly</strong>. Generic add/update/get meta methods should not be used for internal meta data, including "_payment_tokens". Use getters and setters. Backtrace: edit_post, wp_update_post, wp_insert_post, do_action('save_post'), WP_Hook->do_action, WP_Hook->apply_filters, WC_Admin_Meta_Boxes->save_meta_boxes, do_action('woocommerce_process_shop_order_meta'), WP_Hook->do_action, WP_Hook->apply_filters, WC_Meta_Box_Order_Actions::save, do_action('woocommerce_order_action_wcs_process_renewal'), WP_Hook->do_action, WP_Hook->apply_filters, WCS_Admin_Meta_Boxes::process_renewal_action_request, do_action('woocommerce_scheduled_subscription_payment'), WP_Hook->do_action, WP_Hook->apply_filters, WC_Subscriptions_Manager::prepare_renewal, WC_Subscriptions_Manager::process_renewal, wcs_create_renewal_order, wcs_create_order_from_subscription, wcs_copy_order_meta, wcs_set_objects_property, WC_Data->update_meta_data, WC_Data->is_internal_meta_key, wc_doing_it_wrong Plea in /srv/users/usereaea7e7b/apps/usereaea7e7b/public/wp-includes/functions.php on line 5831
[09-Aug-2022 06:32:51 UTC] PHP Notice:  Function is_internal_meta_key was called <strong>incorrectly</strong>. Generic add/update/get meta methods should not be used for internal meta data, including "_payment_tokens". Use getters and setters. Backtrace: edit_post, wp_update_post, wp_insert_post, do_action('save_post'), WP_Hook->do_action, WP_Hook->apply_filters, WC_Admin_Meta_Boxes->save_meta_boxes, do_action('woocommerce_process_shop_order_meta'), WP_Hook->do_action, WP_Hook->apply_filters, WC_Meta_Box_Order_Actions::save, do_action('woocommerce_order_action_wcs_process_renewal'), WP_Hook->do_action, WP_Hook->apply_filters, WCS_Admin_Meta_Boxes::process_renewal_action_request, do_action('woocommerce_scheduled_subscription_payment'), WP_Hook->do_action, WP_Hook->apply_filters, WC_Subscriptions_Manager::prepare_renewal, WC_Subscriptions_Manager::process_renewal, wcs_create_renewal_order, wcs_create_order_from_subscription, wcs_copy_order_meta, wcs_set_objects_property, WC_Data->update_meta_data, WC_Data->add_meta_data, WC_Data->is_internal_meta_key in /srv/users/usereaea7e7b/apps/usereaea7e7b/public/wp-includes/functions.php on line 5831
```

Based on the backtrace in the debug.log, I thought this issue was coming from WooCommerce Subscriptions, but when I tested the same thing with a Stripe subscription I didn't get the same issue.

From the WCPay logs there's an error:

```
2022-08-09T06:24:42+00:00 ERROR There is no saved payment token for order #
```

This issue could be caused by one of these open issues where we're querying the posts table for orders which doesn't exist when COT is enabled:
 - https://github.com/Automattic/woocommerce-payments/issues/4151
 - https://github.com/Automattic/woocommerce-payments/issues/4147

### To Reproduce

1. Enable COT on your store (see [guide](https://github.com/Automattic/woocommerce/wiki/COT-Upgrade-Recipe-Book#how-to-enable-custom-order-tables))
2. Make sure order data is not being synchronized to wp posts table
3. Enable WC Payments and WC Subscriptions
4. Purchase a subscription
5. Go to **WooCommerce > Subscriptions** and go to the Edit Subscription page for the new subscription
6. Using the Subscription actions, choose Process Renewal and then click update.
7. Notice the subscription is now on-hold and the renewal order has failed.

### Actual behavior

Processing a subscription renewal order fails when custom order tables are enabled and syncing is turned off.

### Expected behavior

Processing a subscription renewal when custom order tables is enabled without WP posts syncing enabled, should process the payment successfully and the subscription should flick back to active.

### Desktop (please complete the following information):

* OS: [e.g. iOS]
* Browser [e.g. chrome, safari]
* Version [e.g. 22]

### Smartphone (please complete the following information):

* Device: [e.g. iPhone6]
* OS: [e.g. iOS8.1]
* Browser [e.g. stock browser, safari]
* Version [e.g. 22]

**Comment by mattallan** [comment]
After sitting on this issue a bit longer, I think this might be coming from the Subscriptions copy order meta function not properly copying the payment token from the subscription to the renewal order.

I couldn't reproduce the issue with Stripe hence why I thought it was specific to WCPay but I think that's because Stripe falls back on using customer/user meta to process renewal orders if order meta doesn't exist.

This issue should be address with: https://github.com/Automattic/woocommerce-subscriptions-core/issues/146

**Comment by mattallan** [comment]
This issue was fixed in subscriptions-core 2.5.0 (https://github.com/Automattic/woocommerce-subscriptions-core/pull/229) which was shipped in WC Payments 5.0

All of the remaining HPOS changes were shipped in WC Payments 5.4 therefore we can close this.
